src/librustc_mir/transform/validate.rs

```diff
@@ -38,6 +38,42 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
             &format!("broken MIR in {:?} ({}): {}", self.def_id, self.when, msg.as_ref()),
         );
     }
+
+    /// Check if src can be assigned into dest.
+    /// This is not precise, it will accept some incorrect assignments.
+    fn mir_assign_valid_types(&self, src: Ty<'tcx>, dest: Ty<'tcx>) -> bool {
+        if src == dest {
+            // Equal types, all is good.
+            return true;
+        }
+
+        // Type-changing assignments can happen for (at least) two reasons:
+        // 1. `&mut T` -> `&T` gets optimized from a reborrow to a mere assignment.
+        // 2. Subtyping is used. While all normal lifetimes are erased, higher-ranked types
+        //    with their late-bound lifetimes are still around and can lead to type differences.
+        // Normalize both of them away.
+        // Also see the related but slightly different post-monomorphization method in `interpret/eval_context.rs`.
+        let normalize = |ty: Ty<'tcx>| {
+            ty.fold_with(&mut BottomUpFolder {
+                tcx: self.tcx,
+                // Normalize all references to immutable.
+                ty_op: |ty| match ty.kind {
```

Comment:
what happens when you use `normalize_erasing_regions` here?

Reply:
https://github.com/rust-lang/rust/pull/72796#discussion_r432940863

Reply:
First normalize and then replace mut ref with immut?

Reply:
> But even then, IIRC normalizing_erasing_regions doesn't erase enough lifetimes -- we do need to erase higher-ranked lifetimes bound inside the type here as well due to subtyping.

Reply:
> But even then, IIRC normalizing_erasing_regions doesn't erase enough lifetimes -- we do need to erase higher-ranked lifetimes bound inside the type here as well due to subtyping.

:neutral_face: not quite sure on how to deal with that though

Reply:
I am confused.

That comment explains why, when I wrote this code for Miri, @eddyb told me I could not use `normalizing_erasing_regions`. It's because `normalizing_erasing_regions` doesn't erase *all* regions: regions from `for<'a> ...` quantifications stick around.

Reply:
Afaict you have 2 problems rn:
- we need to normalize opaque types, projections and constants
- we need to erase all lifetimes

While I am not certain if this is the right way, I think this could be solved by
first using `normalize_erasing_regions`, which should fix `1` without completely solving `2`.
I believe your current approach already solves 2.

*editorial note: @lcnr has only very specific knowledge and may or may not actually know what he is talking about here*

Reply:
That didn't do it, unfortunately. :/